pkg/issuer/vault/issue.go

```diff
@@ -0,0 +1,200 @@
+		return nil, nil, err
+	}
+
+	crt.UpdateStatusCondition(v1alpha1.CertificateConditionReady, v1alpha1.ConditionTrue, successCertIssued, messageCertIssued)
+
+	return key, certPem, nil
+}
+
+func (v *Vault) obtainCertificate(ctx context.Context, crt *v1alpha1.Certificate) ([]byte, []byte, error) {
+	// get existing certificate private key
+	signeeKey, err := kube.SecretTLSKey(v.secretsLister, crt.Namespace, crt.Spec.SecretName)
+	if k8sErrors.IsNotFound(err) || errors.IsInvalidData(err) {
+		signeeKey, err = pki.GenerateRSAPrivateKey(keyBitSize)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error generating private key: %s", err.Error())
+		}
+	}
+
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting certificate private key: %s", err.Error())
+	}
+
+	commonName := crt.Spec.CommonName
+	altNames := crt.Spec.DNSNames
+	if len(commonName) == 0 && len(altNames) == 0 {
+		return nil, nil, fmt.Errorf("no domains specified on certificate")
+	}
+
+	crtPem, err := v.signCertificate(crt, signeeKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pki.EncodePKCS1PrivateKey(signeeKey), crtPem, nil
+}
+
+// signCertificate returns a signed x509.Certificate object for the given
+// *v1alpha1.Certificate crt.
+func (v *Vault) signCertificate(crt *v1alpha1.Certificate, key *rsa.PrivateKey) ([]byte, error) {
+	commonName, err := pki.CommonNameForCertificate(crt)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting Common Name from certificate: %s", err.Error())
+	}
+
+	altNames, err := pki.DNSNamesForCertificate(crt)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting DNS Names from certificate: %s", err.Error())
+	}
+
+	if len(commonName) == 0 && len(altNames) > 0 {
+		commonName = altNames[0]
+	}
+
+	template := pki.GenerateCSR(commonName, altNames...)
+
+	derBytes, err := x509.CreateCertificateRequest(rand.Reader, template, key)
+	if err != nil {
+		return nil, fmt.Errorf("error creating x509 certificate: %s", err.Error())
+	}
+
+	pemRequestBuf := &bytes.Buffer{}
+	err = pem.Encode(pemRequestBuf, &pem.Block{Type: "CERTIFICATE REQUEST", Bytes: derBytes})
+	if err != nil {
+		return nil, fmt.Errorf("error encoding certificate request: %s", err.Error())
+	}
+
+	return v.requestVaultCert(commonName, altNames, pemRequestBuf.String())
+}
+
+func (v *Vault) requestVaultCert(commonName string, altNames []string, csr string) ([]byte, error) {
+	client, err := vault.NewClient(nil)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing Vault client: %s", err.Error())
+	}
+
+	client.SetAddress(v.issuer.GetSpec().Vault.Server)
+
+	tokenRef := v.issuer.GetSpec().Vault.Auth.TokenSecretRef
+	token, err := vaultTokenRef(v.secretsLister, v.issuerResourcesNamespace, tokenRef.Name, "token")
+	if err != nil {
+		return nil, fmt.Errorf("error reading Vault token from secret %s/%s: %s", v.issuerResourcesNamespace, tokenRef.Name, err.Error())
+	}
```

Comment:
We should probably check here to see *why* it failed (via the `k8s.io/apimachinery/pkg/api/errors` package).

This isn't something that's handled well in the CA or ACME implementation currently, and I think would take some fairly significant refactoring to add retrospectively (as the top level sync function needs a way to return nil despite an error occurring).

When initially writing cert-manager, I created a `TransientError` type that could be used to indicate a sync should be retried. This would save all the refactoring, at the expense of needing to check for transient errors at various points (and/or not 'masking' errors with `fmt.Errorf`)